java/src/main/java/ai/rapids/cudf/TableWithMeta.java

```diff
@@ -0,0 +1,67 @@
+/*
+ *
+ *  Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+
+package ai.rapids.cudf;
+
+/**
+ * A table along with some metadata about the table. This is typically returned when
+ * reading data from an input file where the metadata can be important.
+ */
+public class TableWithMeta implements AutoCloseable {
```

Comment:
Can this be implemented as a derived class extending `Table`?

Reply:
It would not match what the C++ code is doing. If you want me to try I can. I agree that conceptually it would be more interesting, but I wanted to preserve the C++ API in this case. Also it gets to be a more invasive change to make it work.

Reply:
I see. So it is just some wrapper for special purposes.

Reply:
Yes. For all of the other APIs where we read the from a file we already know the names of the columns, as we passed them into the reader. So in those cases we just return a Table with the columns in the order that was requested when it was passed in.  Here we need the metadata to be able to know what the names of the columns are.  Long term we should probably update all of our reader APIs to match what CUDF is doing in C++ and just return the `TableWithMeta`. But for now to minimize the impact of the change we decided not to do that. I do like the idea of having `TableWithMeta` be a `Table` too. We probably will/should implement that if we do try to update all of the other APIs, just because it would minimize the impact of the change on customer code.